src/server/zset_family.cc

```diff
@@ -870,61 +898,187 @@ OpResult<AddResult> OpAdd(const OpArgs& op_args, const ZParams& zparams, string_
   return aresult;
 }
 
-struct StoreArgs {
+struct SetOpArgs {
   AggType agg_type = AggType::SUM;
   unsigned num_keys;
   vector<double> weights;
+  bool with_scores = false;
 };
 
-OpResult<StoreArgs> ParseStoreArgs(CmdArgList args) {
-  string_view num_str = ArgS(args, 2);
-  StoreArgs store_args;
+OpResult<void> FillAggType(string_view agg, SetOpArgs* op_args) {
+  if (agg == "SUM") {
+    op_args->agg_type = AggType::SUM;
+  } else if (agg == "MIN") {
+    op_args->agg_type = AggType::MIN;
+  } else if (agg == "MAX") {
+    op_args->agg_type = AggType::MAX;
+  } else {
+    return OpStatus::SYNTAX_ERR;
+  }
+  return OpStatus::OK;
+}
+
+OpResult<unsigned> ParseAggregate(CmdArgList args, bool store, SetOpArgs* op_args) {
+  if (store && args.size() != 2) {
+    return OpStatus::SYNTAX_ERR;
+  }
+  if (!store && args.size() < 2) {
+    return OpStatus::SYNTAX_ERR;
+  }
 
+  ToUpper(&args[1]);
+  const auto filled = FillAggType(ArgS(args, 1), op_args);
+  if (!filled) {
+    return filled.status();
+  }
+  return 1;
+}
+
+OpResult<unsigned> ParseWeights(CmdArgList args, SetOpArgs* op_args) {
+  if (args.size() <= op_args->num_keys) {
+    return OpStatus::SYNTAX_ERR;
+  }
+
+  for (unsigned i = 0; i < op_args->num_keys; ++i) {
+    string_view weight = ArgS(args, i + 1);
+    if (!absl::SimpleAtod(weight, &(op_args->weights[i]))) {
+      return OpStatus::INVALID_FLOAT;
+    }
+  }
+  return op_args->num_keys;
+}
+
+OpResult<unsigned> ValidateKeyCount(string_view arg_num_keys) {
   // we parsed the structure before, when transaction has been initialized.
-  CHECK(absl::SimpleAtoi(num_str, &store_args.num_keys));
-  DCHECK_GE(args.size(), 3 + store_args.num_keys);
+  unsigned num_keys;
+  if (!absl::SimpleAtoi(arg_num_keys, &num_keys)) {
+    return OpStatus::SYNTAX_ERR;
+  }
+  return num_keys;
+}
+
+OpResult<unsigned> ParseWithScores(CmdArgList args, SetOpArgs* op_args) {
+  if (args.size() != 1) {
+    return OpStatus::SYNTAX_ERR;
+  }
+  op_args->with_scores = true;
+  return 0;
+}
+
+OpResult<SetOpArgs> ParseSetOpArgs(CmdArgList args, bool store) {
+  // TODO: support variadic key for ZUNION command (now fixed to 3)
+  string_view num_str = store ? ArgS(args, 2) : "3";
+  SetOpArgs op_args;
 
-  store_args.weights.resize(store_args.num_keys, 1);
-  for (size_t i = 3 + store_args.num_keys; i < args.size(); ++i) {
+  const auto parsed = ValidateKeyCount(num_str);
+  if (!parsed) {
+    return parsed.status();
+  }
+  op_args.num_keys = *parsed;
+  unsigned key_start = store ? 3 + op_args.num_keys : 1 + op_args.num_keys;
```

Comment:
I would call it opt_args_start if this implies to at which index the optional params are in args list